### PR TITLE
Change read_ini.sh to use associative arrays

### PR DESCRIPTION
**Pull request by jcass77**
- Allows special characters to be used in section names of settings.ini.

- Adds support for configuring extensions like [local-sqlite] properly.

- Note that support for the ```search_limit``` parameter requires Mopidy-Local-SQLite >= 0.9.2 (run ```pip install --upgrade Mopidy-Local-SQLite``` to upgrade)

### CODE REVIEW

File: filechanges/boot/config/settings.ini [hunk 1]
```diff
@@ -331,13 +330,42 @@ playlists_dir = /var/lib/mopidy/playlists
...
 #You can set it to json or whoosh. Whoosh might be better with large collections, JSON is a bit more mature.
-# TODO: You cannot add settings for [local-sqlite]
-# because it breaks the startup script (won't read dashes in section names
 library = sqlite
 #library = whoosh
 #library = json
 
 scan_timeout = 1000
-scan_flush_threshold = 1000
+scan_flush_threshold = 100
```
Comment (kingosticks) [inline filechanges/boot/config/settings.ini]: Is this change intentional?

**Comment by kingosticks** [comment]
I've tried and failed to review this on github as their diff tool has made a real mess of things. I'll try to test it out tomorrow but from a quick look it seems this will now accept empty section names. Probably not the end of the world.

File: filechanges/boot/config/settings.ini (hunk 1 above)
Comment (jcass77) [inline filechanges/boot/config/settings.ini]: Yes the change to ```scan_flush_threshold = 100``` is intentional and to align with the configuration recommendations as listed at https://github.com/tkem/mopidy-local-sqlite

**Comment by jcass77** [comment]
Most of the diffs are just as a result of leading space changes on each line (I didn't feel up to trying to match it all by hand). It may be easier to just compare the 'master' and 'head' branches of ```read_ini.sh``` on https://github.com/rudimeier/bash_ini_parser as the files are direct copies from that repository.

File: filechanges/opt/musicbox/read_ini.sh [hunk 2]
```diff
@@ -11,281 +11,277 @@
...
-	if [[ "${line}" =~ ^\[[a-zA-Z0-9_]{1,}\]$ ]]
-	then
+		# Section marker?
+		if [[ "${line}" =~ ^\[.*\]$ ]]
+		then
+			# Set SECTION var to name of section (strip [ and ] from section marker)
+			SECTION="${line#[}"
+			SECTION="${SECTION%]}"
+			read SECTION <<<"${SECTION}"
+			if ! [[ "${line}" =~ ^[[:print:]]+$  ]] ;then
```
Comment (jcass77) [inline filechanges/opt/musicbox/read_ini.sh]: If we want to avoid empty section markers then we probably need to replace ```^[[:print:]]+$``` with something like ```^\[.*\S.*\]+$```. But I haven't tested this as I didn't want to fiddle with the original script.

Comment (kingosticks) [inline filechanges/opt/musicbox/read_ini.sh]: OK, I thought that's what you had changed but I must just have misread the
diff.

File: filechanges/opt/musicbox/read_ini.sh [hunk 3]
```diff
@@ -11,281 +11,277 @@
...
+				continue
+			fi
+		fi
 
-	    continue
-	fi
+		# Valid var/value line? (check for variable name and then '=')
+		# HACK: Handling of local-sqlite's 'directories' variables. Bypass lines ending in 'directory:'
+		#       as if they are valid.
+		if ! [[ "${line}" =~ ^[[:print:]]+[[:space:]]*directory:|= ]]
```
Comment (kingosticks) [inline filechanges/opt/musicbox/read_ini.sh]: Urgh, this isn't pleasant. But maybe you can me understand, am I right that this matches a line not starting with 1 or more printable chars, followed by 0 or more spaces and then either the literal string 'directory:' OR an equals character?

And that's supposed to match something like the following?
'Genres                  local:directory?type=genre' ?

Does it?

Comment (kingosticks) [inline filechanges/opt/musicbox/read_ini.sh]: I get this error:
> pair: 'local-sqlite__Albums local:directory?type' = 'album'
/opt/musicbox/read_ini.sh: line 276: local-sqlite__Albums local:directory?type: syntax error in expression (error token is "local:directory?type")

Comment (jcass77) [inline filechanges/opt/musicbox/read_ini.sh]: There's a comment in https://github.com/rudimeier/bash_ini_parser/issues/2 explaining the specific scenario that this hack is meant to work around. The local-sqlite 'directories' parameter is unusually complex in that it consists of multiple lines of values, each potentially also containing an '=' sign:

```
# top-level directories for browsing, as <name> <uri>
directories =
    Albums                  local:directory?type=album
    Artists                 local:directory?type=artist
    Composers               local:directory?type=artist&role=composer
    Folders                 local:directory:
    Genres                  local:directory?type=genre
    Performers              local:directory?type=artist&role=performer
    Release Years           local:directory?type=date&format=%25Y
    Tracks                  local:directory?type=track
    Last Week's Updates     local:directory?max-age=604800
    Last Month's Updates    local:directory?max-age=2592000
```

I'm not sure how to process something like that in bash and still get an array of valid key/value pairs. So for now, we just want the script to run through as we won't use these specific variables in any of the Pi Musicbox scripts. The hack in the regular expression is just supposed to handle this line:

```Folders                 local:directory:```

...because the line does not contain an '=' sign it is not seen as a valid key/value pair and the script fails. By adding ```directory:``` to the regular expression, lines ending in 'directory:` will be parsed as well just to handle this one exception.

Comment (jcass77) [inline filechanges/opt/musicbox/read_ini.sh]: Did you perhaps encounter the error when testing the script by running ```setsound.sh```?

There was a problem with the ini parsing routines in ```setsound.sh``` not being in sync with ```startup.sh```, so if you tried to run the script manually without going through a reboot it would fail. I've fixed this in https://github.com/jcass77/Pi-MusicBox/commit/bb7032bf09d88efdd341941872138b62da4bdb38

Comment (kingosticks) [inline filechanges/opt/musicbox/read_ini.sh]: Yes, I was running it from setsound.sh. For some reason that I cannot explain the script wouldn't work when trying to call it directly (or at all on my Mint machine) so I just settled for that since it at least sort of worked. I hate bash. 

Let's at least remove that `directories` setting from the file anyway, I'm going to blindly assume it's the default value so we don't need to explicitly set it. And it's making a long file even longer.

Comment (jcass77) [inline filechanges/opt/musicbox/read_ini.sh]: Ok, I've reverted to the original implementation of read_ini.sh and commented out the offending lines in settings.ini - please check if this works for you as well?

File: filechanges/boot/config/settings.ini [hunk 4]
```diff
@@ -331,13 +330,46 @@ playlists_dir = /var/lib/mopidy/playlists
...
 excluded_file_extensions = .html, .jpeg, .jpg, .log, .nfo, .png, .txt, .mkv, .avi, .divx, .qt, .wmv, .htm, .zip, .rar, .gz, .pdf, .exe, .ini, .mid, .db, .m3u, .sfv, .midi
 
+[local-sqlite]
+enabled = true
+
+# NOTE: Handling of complex parameters (e.g. that contain spaces and do not contain '=') are not
```
Comment (jcass77) [inline filechanges/boot/config/settings.ini]: I've added a note and commented out the lines that cannot currently be parsed by read_ini.sh

**Comment by jcass77** [comment]
Rebasing this on /upstream/develop to make merging and testing easier. See https://github.com/woutervanwijk/Pi-MusicBox/pull/253
